### PR TITLE
Add joystick vibration support on Linux

### PR DESCRIPTION
**Pull request by wbrbr**
Related to #5013 
This pull request adds joystick vibration support on Linux.
Here is the cross platform API:
- `Input.start_joy_vibration(device, weak_magnitude, strong_magnitude)` to start the joystick vibration
- `Input.stop_joy_vibration(device)` to stop the vibration
- `Input.get_joy_vibration(device)` to get the current strength of the vibration

**TODO**:
- Add support for Windows and OS X (unfortunately I don't know the API for these platforms, hopefully someone could add support after this PR is merged)
- Update the joystick demo (I will do that myself when this PR is merged)

I don't know much about Godot's codebase, so don't hesitate to tell me if I did something wrong :)

### CODE REVIEW

**Comment by akien-mga** [comment]
Ping our input experts @Hinsbart and @punto- :)

**Comment by volzhs** [comment]
We can add vibration for Android and iOS later.
It would be nice that we can have more general method name for vibration.

**Comment by nounoursheureux** [comment]
@volhz That's a good point. I have a proposition for  a more general way of handling vibration:
We could have a `Vibration`/`Haptic` singleton, with the following methods:
- `Vibration.start(device, strength)`
- `Vibration.stop(device)`
- `Vibration.get_strength(device)`
- `Vibration.get_default_device()`: returns the device corresponding to the smartphone or the first joystick, useful for single player games
- `Vibration.get_device_from_joystick(joystick)`: return the haptic device corresponding to the joystick

Pros:
- More general approach (can support both smartphones and joysticks)
- Vibration is not really input, so it doesn't make sense to have it as a part of the `Input` class

Cons:
- Joysticks usually have two rumble motors, a strong one and a weak one. With this approach, the developer would have less control over the individual motors.
- May be more complicated to implemented

Do you think this would be better ?

**Comment by volzhs** [comment]
@nounoursheureux sounds good. I also agree with **Vibration is not really input**.

**Comment by Hinsbart** [comment]
Oh wow, nice.
Any reason why you're looping through /dev/input again when looking for FF capabilities, instead of using the existing fd?
Apart from that, it looks pretty good.
Will give it a thorough check the next days, I'm really sorry for the work you did there (I've had code almost like this for a few months now) :s

Anyway, I propose adding a `duration` parameter (takes time in seconds, after which to automatically stop the vibration).

**Comment by nounoursheureux** [comment]
@Hinsbart 
> Any reason why you're looping through /dev/input again when looking for FF capabilities, instead of using the existing fd?

If by "the existing fd" you mean the one pointing to */dev/input/js0*, that's because the joystick device and the force feedback device are not the same (at least with my Xbox 360 controller), so I can't use force feedback ioctls on */dev/input/js0*.

> I'm really sorry for the work you did there (I've had code almost like this for a few months now) :s

No problem, that was a great way to learn about both Linux and Godot :)

> Anyway, I propose adding a duration parameter (takes time in seconds, after which to automatically stop the vibration).

I thought about that, and it would be easy for Linux since the driver already manages this, but XInput doesn't AFAIK so I think it would be better to let the users manage this with *Timer* nodes.

Also, what do you think about my alternative API proposition ?

**Comment by volzhs** [comment]
> Anyway, I propose adding a duration parameter (takes time in seconds, after which to automatically stop the vibration).

I like this too.

**Comment by Hinsbart** [comment]
Mhh, on my 360 controllers (all: wired, wireless and 3rd party clone) force feedback is always on the same device as the reading events... Maybe you forgot to change the mode to O_RDWR? ^ ^

> I thought about that, and it would be easy for Linux since the driver already manages this, but XInput doesn't AFAIK so I think it would be better to let the users manage this with Timer nodes.

Don't worry about Xinput, already got it :) (https://github.com/Hinsbart/godot/commit/43c3938f233ac122e66711f38047ff2c94687f7b)
(tho I'll have to adapt it to your api a bit)

About the api proposition, should be good I guess.

**Comment by punto-** [comment]
You can keep both parameters for the vibration, if the device doesn't
support a second motor just ignore the value..

**Comment by nounoursheureux** [comment]
> Mhh, on my 360 controllers (all: wired, wireless and 3rd party clone) force feedback is always on the same device as the reading events... Maybe you forgot to change the mode to O_RDWR?

Aaaaaaand you were right. Fixed it.

**Comment by reduz** [comment]
@punto- merge this when you feel it's ready

**Comment by reduz** [comment]
would like to see it implemented on windows at least (I don't think OSX supports joystick does it?)

**Comment by Hinsbart** [comment]
@nounoursheureux you forgot to remove the `dir_filter()` function :p 
I think you may also rebase this branch on that occasion.
Apart from that it's good and ready for merge I guess :)
I'll probably make some adjustments to it but I think it's easier when it get's merged first.

**Comment by nounoursheureux** [comment]
@Hinsbart I am adding a duration parameter, I will rebase when this is done.

**Comment by Hinsbart** [comment]
Cool :)

**Comment by nounoursheureux** [comment]
@Hinsbart Done :). Tell me if you see something wrong !

**Comment by Hinsbart** [comment]
> Tell me if you see something wrong !

Almost perfect. Only thing I spotted is that we could mention in the docs that a duration of 0 will play the vibration indefinitely (or as long as the device is able to).
Also maybe 0 could be the default value for duration. Idk, that's up for discussion ^^

File: core/os/input.cpp [hunk 1]
```diff
@@ -59,6 +59,10 @@ void Input::_bind_methods() {
 	ObjectTypeDB::bind_method(_MD("get_joy_axis","device","axis"),&Input::get_joy_axis);
 	ObjectTypeDB::bind_method(_MD("get_joy_name","device"),&Input::get_joy_name);
 	ObjectTypeDB::bind_method(_MD("get_joy_guid","device"),&Input::get_joy_guid);
+	ObjectTypeDB::bind_method(_MD("get_joy_vibration_strength", "device"), &Input::get_joy_vibration_strength);
+	ObjectTypeDB::bind_method(_MD("get_joy_vibration_duration", "device"), &Input::get_joy_vibration_duration);
+	ObjectTypeDB::bind_method(_MD("start_joy_vibration", "device", "strong_magnitude", "weak_magnitude", "duration"), &Input::start_joy_vibration);
```
Comment (Hinsbart) [inline core/os/input.cpp]: the magnitude parameters are switched here, weak is the first one :p

**Comment by nounoursheureux** [comment]
Fixed it :P

**Comment by nounoursheureux** [comment]
I also updated the docs.

**Comment by punto-** [comment]
Is this the one with the methods on Input, or the Vibration class?

Probably for another PR: we could have a Node that exposes these parameters
for a configured device, then you can include vibration effects in
animations

**Comment by akien-mga** [comment]
This one extends Input. But I'm pretty sure @nounoursheureux or @Hinsbart will be happy to move it to a better class once there's a consensus about what form it should have.

**Pull request merged**